### PR TITLE
test: refactor and stabilize e2e tests

### PR DESCRIPTION
**Pull request by Seamoo13**
#### Please provide a brief summary of this pull request.
https://github.com/SAP/fundamental-ngx/issues/4166

refactored:
- Fixed-Card-Layout 
- Link 

updated tests:
- List
- Standard-list-item
- Object-list-item

Stabilization:
- added fixes for page loading
- added automatic retries for before and afterEach hooks
- added automatic retries for when the whole spec failed (timeout or connection issues)

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

File: e2e/wdio/platform/tests/link.e2e-spec.ts [hunk 1]
```diff
@@ -82,76 +74,70 @@ xdescribe('Link component test suite', function() {
...
+        return;
+    }
+    console.log('skip hover check for IE, Safari');
+}
 
-function checkLinkTargetDestination(element, site: string): void {
+function checkLinkTarget(element, site: string): void {
     click(element);
-    expect(browser).toHaveUrlContaining(site);
+    pause(5000);
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/platform/tests/link.e2e-spec.ts:134]: maybe use waitUntil with browser.getUrl() to see if url is changed?

File: e2e/wdio/platform/tests/list.e2e-spec.ts [hunk 2]
```diff
@@ -150,7 +150,7 @@ describe('List test suite:', function() {
 
         it('should check scroll', () => {
             // skip for FF due to issue https://github.com/SAP/fundamental-ngx/issues/4107
-            if (!browserIsFirefox()) {
+            if (!browserIsFirefox() && !browserIsSafari()) {
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/platform/tests/list.e2e-spec.ts:153]: same as above for skip

**Review by AntonOlkhovskyi — CHANGES_REQUESTED** [review]

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts [hunk 3]
```diff
@@ -71,90 +57,83 @@ describe('Fixed card layout test suite', function() {
...
-                const newFirstCardText = getText(fxdCardLayoutPg.cardDivArr);
-                const newSwapCardText = getText(fxdCardLayoutPg.cardDivArr, 4);
-                expect(newFirstCardText).not.toBe(originalFirstCardText);
-                expect(newSwapCardText).not.toBe(originalSwapCardText);
+            if (browserIsIE()) {
+                console.log('skip IE because of #3882');
                 return;
             }
-            console.log('Firefox for Safari and IE');
+            const originalFirstCardText = getText(fxdCardLayoutPg.cardDivArr, 0);
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts:92]: remove 0 index

File: e2e/wdio/platform/tests/list.e2e-spec.ts (hunk 2 above)
Comment (AntonOlkhovskyi) [reply e2e/wdio/platform/tests/list.e2e-spec.ts:153]: fix if.

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts [hunk 4]
```diff
@@ -178,50 +157,30 @@ describe('Fixed card layout test suite', function() {
...
-                expect([null, '']).toContain(dirValueBefore);
-                click(fxdCardLayoutPg.rtlSwitcherArr, i);
-                expect(getCSSPropertyByName(fxdCardLayoutPg.exampleAreaContainersArr, 'direction', i).value).toBe('rtl');
-                expect(getAttributeByName(fxdCardLayoutPg.exampleAreaContainersArr, 'dir', i)).toBe('rtl');
-            }
-        });
     });
 
-    function checkDragAndDrop(clickElement, startLocation, endElementLocation): any {
+    function checkDragAndDrop(clickElement, startLocation, endLocation, endLocationIndex): any {
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts:176]: why any ?

File: e2e/wdio/platform/pages/checkbox.po.ts [hunk 5]
```diff
@@ -27,6 +28,7 @@ export class CheckboxPO extends BaseComponentPo {
     rtlSwitcherArr = 'rtl-switch .fd-switch__handle';
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/platform/pages/checkbox.po.ts:28]: can be removed, it is now in base-component

File: e2e/wdio/platform/tests/list.e2e-spec.ts [hunk 6]
```diff
@@ -141,7 +141,7 @@ describe('List test suite:', function() {
 
     describe('Virtual Scroll examples:', function() {
         it('should do basic checks', () => {
-            checkElArrIsClickable(listPg.vScrollListItems);
+            isElementClickable(listPg.vScrollListItems, 0);
```
Comment (AntonOlkhovskyi) [inline e2e/wdio/platform/tests/list.e2e-spec.ts:144]: remove 0

**Review by AntonOlkhovskyi — CHANGES_REQUESTED** [review]

File: e2e/wdio/platform/pages/checkbox.po.ts (hunk 5 above)
Comment (nikvalor) [reply e2e/wdio/platform/pages/checkbox.po.ts:28]: @Seamoo13 Can you remove rtlSwitcherArr = 'rtl-switch .fd-switch__handle'; ?

File: e2e/wdio/platform/pages/display-list-item.po.ts [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
+    cozyDisplayTitles = 'fdp-borderless-display-list-item-example fdp-list:first-of-type a span';
+    comfyDisplayTitles = 'fdp-borderless-display-list-item-example fdp-list:nth-of-type(2) a span';
     sections = 'fdp-borderless-display-list-item-example ul';
 
     declarativeDisplayLinks = 'fdp-display-list-item-with-navigation-example a';
-    declarativeDisplayTitles = 'fdp-display-list-item-with-navigation-example fdp-list:first-of-type span:first-of-type';
+    declarativeDisplayTitles = 'fdp-display-list-item-with-navigation-example a span:first-of-type';
     declarativeSection = 'fdp-display-list-item-with-navigation-example ul';
 
     exampleAreaContainersArr = '.fd-doc-component';
```
Comment (nikvalor) [inline e2e/wdio/platform/pages/display-list-item.po.ts:18]: Please remove exampleAreaContainersArr = '.fd-doc-component';

File: e2e/wdio/platform/tests/list.e2e-spec.ts [hunk 8]
```diff
@@ -193,17 +192,21 @@ describe('List test suite:', function() {
 
         it('should check delete action', () => {
             click(listPg.btnDeleteBtn, 0);
```
Comment (nikvalor) [inline e2e/wdio/platform/tests/list.e2e-spec.ts:194]: remove 0

File: e2e/wdio/platform/tests/list.e2e-spec.ts [hunk 9]
```diff
@@ -193,17 +192,21 @@ describe('List test suite:', function() {
...
+            if (browserIsIE()) {
+                acceptAlert();
+                return;
             }
+            expect(getAlertText()).toContain('Delete row');
             acceptAlert();
         });
 
         it('should check edit action', () => {
             click(listPg.btnEditBtn, 0);
```
Comment (nikvalor) [inline e2e/wdio/platform/tests/list.e2e-spec.ts:204]: remove 0

Comment (Seamoo13) [reply e2e/wdio/platform/tests/list.e2e-spec.ts]: done

File: e2e/wdio/platform/tests/list.e2e-spec.ts (hunk 8 above)
Comment (Seamoo13) [reply e2e/wdio/platform/tests/list.e2e-spec.ts]: done

File: e2e/wdio/platform/pages/display-list-item.po.ts (hunk 7 above)
Comment (Seamoo13) [reply e2e/wdio/platform/pages/display-list-item.po.ts]: done

File: e2e/wdio/platform/tests/list.e2e-spec.ts (hunk 6 above)
Comment (Seamoo13) [reply e2e/wdio/platform/tests/list.e2e-spec.ts]: done

File: e2e/wdio/platform/pages/checkbox.po.ts (hunk 5 above)
Comment (Seamoo13) [reply e2e/wdio/platform/pages/checkbox.po.ts]: done

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts (hunk 4 above)
Comment (Seamoo13) [reply e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts]: updated

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts (hunk 3 above)
Comment (Seamoo13) [reply e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts]: done

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts [hunk 10]
```diff
@@ -178,50 +157,30 @@ describe('Fixed card layout test suite', function() {
...
-            const clickElement = elementArray(fxdCardLayoutPg.disableDragBtn);
-            const cards = elementArray(fxdCardLayoutPg.cardDivArr);
+            const originalFirstCardText = getText(fxdCardLayoutPg.disabledCardDiv, 0);
 
             scrollIntoView(fxdCardLayoutPg.disableDragBtn);
             click(fxdCardLayoutPg.disableDragBtn);
-            checkDragAndDrop(clickElement[0], cards[9], cards[13]);
-            const newFirstCardText = getText(fxdCardLayoutPg.cardDivArr, 9);
+            checkDragAndDrop(fxdCardLayoutPg.disabledCardContent, fxdCardLayoutPg.disabledCardDiv, fxdCardLayoutPg.disabledCardContent, 4);
+            const newFirstCardText = getText(fxdCardLayoutPg.cardDivArr, 0);
```
[the remark this answers is not in the document]
Comment (Seamoo13) [reply e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts]: done

File: e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts [hunk 11]
```diff
@@ -178,50 +157,30 @@ describe('Fixed card layout test suite', function() {
         });
 
         it('should check drag and drop is disabled', () => {
-            const originalFirstCardText = getText(fxdCardLayoutPg.cardDivArr, 9);
-            const clickElement = elementArray(fxdCardLayoutPg.disableDragBtn);
-            const cards = elementArray(fxdCardLayoutPg.cardDivArr);
+            const originalFirstCardText = getText(fxdCardLayoutPg.disabledCardDiv, 0);
```
[the remark this answers is not in the document]
Comment (Seamoo13) [reply e2e/wdio/core/tests/fixed-card-layout.e2e-spec.ts]: done

**Review by AntonOlkhovskyi — APPROVED** [review]

**Review by nikvalor — APPROVED** [review]

File: e2e/wdio/driver/wdio.ts [hunk 12]
```diff
@@ -39,6 +39,15 @@ export function browserIsSafari(): boolean {
     return isBrowser('Safari');
 }
 
+export function browserIsSafariorFF(): boolean {
+    if (browserIsSafari()) {
+        return true;
+    } else if (browserIsFirefox()) {
```
Comment (valorkin) [inline e2e/wdio/driver/wdio.ts:45]: please remove `else`

**Review by valorkin — APPROVED** [review]
lgtm

**Pull request merged**